File: solutions/day14.ts
import fs from "node:fs";

const raw = fs.readFileSync("./input.txt", "utf-8");
const robots = raw.split("\n").map((robot) => robot.split(" "));
const [locations, velocities] = robots[0]
  .map((_, colIndex) => robots.map((row) => row[colIndex]))
  .map((items) =>
    items.map((item) => item.split("=")[1].split(",").map(Number))
  );

function multiplyQuadrants(
  locations: number[][],
  dimensions: [number, number]
): number {
  const [width, height] = dimensions.map((dim) => Math.floor(dim / 2));
  const quadrants = locations.reduce(
    (acc, location) => {
      if (location[0] > width && location[1] < height) {
        acc[0]++;
      } else if (location[0] < width && location[1] < height) {
        acc[1]++;
      } else if (location[0] < width && location[1] > height) {
        acc[2]++;
      } else if (location[0] > width && location[1] > height) {
        acc[3]++;
      }
      return acc;
    },
    [0, 0, 0, 0]
  );
  //   console.log(quadrants);
  return quadrants.reduce((acc, quadrant) => acc * quadrant, 1);
}

function displayLocations(
  locations: number[][],
  dimensions: [number, number]
): string {
  const [width, height] = dimensions;
  const grid = Array.from({ length: height }, () =>
    Array.from({ length: width }, () => ".")
  );
  locations.forEach((location) => {
    grid[location[1]][location[0]] = "#";
  });
  const output = grid.map((row) => row.join("")).join("\n");
  return output;
}

function mod(n: number, m: number): number {
  return ((n % m) + m) % m;
}

const width = 101;
const height = 103;
let i = 1;
let resultsPart1 = 0;
while (true) {
  locations.forEach((location, j) => {
    locations[j][0] = mod(location[0] + velocities[j][0], width);
    locations[j][1] = mod(location[1] + velocities[j][1], height);
  });
  if (i === 100) {
    resultsPart1 = multiplyQuadrants(locations, [width, height]);
  }
  const display = displayLocations(locations, [width, height]);
  if (display.includes(new Array(10).fill("#").join(""))) {
    console.log(display);
    console.log("Part 1:", resultsPart1);
    console.log("Part 2:", i);
    break;
  }
  i++;
}

File: solutions/python/day14.py
with open("./input.txt", "r") as txt:
  rob: str = txt.read().rstrip()
  txt.close()

robots = list(map(list, zip(*[robot.split(' ') for robot in rob.split('\n')])))
locations = [list(map(int, location.split('=')[-1].split(','))) for location in robots[0]]
velocities = [list(map(int, velocity.split('=')[-1].split(','))) for velocity in robots[1]]

def multiplyQuadrants(locations: list[list[int]], width: int, height: int) -> int:
    width //= 2
    height //= 2
    quadrant = [0] * 4
    for location in locations:
        if location[0] < width and location[1] < height:
            quadrant[0] += 1
        elif location[0] > width and location[1] < height:
            quadrant[1] += 1
        elif location[0] < width and location[1] > height:
            quadrant[2] += 1
        elif location[0] > width and location[1] > height:
            quadrant[3] += 1
    # print(quadrant)
    return quadrant[0] * quadrant[1] * quadrant[2] * quadrant[3]

def getLocationsDisplay(locations: list[list[int]], width: int, height: int) -> str:
    grid = [['.'] * width for _ in range(height)]
    for location in locations:
        grid[location[1]][location[0]] = '#'
    result = ''
    for row in grid:
        result += ''.join(row) + '\n'
    return result[:-1]

width = 101
height = 103
i = 1
while True:
    for j in range(len(locations)):
        locations[j][0] = (locations[j][0] + velocities[j][0]) % width
        locations[j][1] = (locations[j][1] + velocities[j][1]) % height
    if i == 100:
        resultsPart1 = multiplyQuadrants(locations, width, height)
    display = getLocationsDisplay(locations, width, height)
    if '#' * 10 in display:
        print(display)
        print("Part 1:", resultsPart1)
        print("Part 2:", i)
        break
    i += 1
